### ISSUE TITLE
Add a mechanism to migrate the database

### ISSUE THREAD

**Issue by brianhelba** [girder/girder#946]
As Improvements and bugfixes continue to be made to the Model layer, the database may occasionally need to be updated too. Rather than have lots of inline checks for legacy database formats, it would be nice to have a system for one-time migration operations that run against the database.

This differs from the existing system consistency check in that migration would address very specific issues that we know need to be changed, would be run only once per issue, and would be mandatory to run before a new version of Girder is deployed.

To enable this, we may need to add a database version field to the database (perhaps as a hidden Setting). When Girder starts up, it will check the database version against the current value in the migration mechanism; if the database is older than the latest migration, than all migrations between the database version and the code version are run once.

For safety, each migration should be idempotent, but in theory they will only be run once. Additionally, migrations should generally be written using raw Pymongo operations, since we don't know which version of the model layer will be present when a migration is finally run.

It is possible that for large databases, migrations will take some time, so the proposed mechanism of running them automatically on startup will need to be well-documented, with migrations only being added on minor releases (major releases is probably too slow, since migrations could be bugfixes, that don't come with a new API). Alternatively, we could try to follow Django's typical workflow ( https://docs.djangoproject.com/en/1.8/topics/migrations/ ) and run them manually, but this would necessitate a lot more compatibility code and defeat a large part of the reason for having migrations.

**Comment by brianhelba** [comment]
Some sample PRs which should have accompanying migrations:
* https://github.com/girder/girder/pull/810
* https://github.com/girder/girder/pull/945

**Comment by zachmullen** [comment]
:+1: , this should be pretty straightforward to achieve.

**Comment by sebastienbarre** [comment]
Feel free to hit me up if you decide to discuss it in a meeting. We have migrated the KWiK database schema 226 (!) times as of today, and most often live during the day while you guys are using the app.
